### PR TITLE
Add support for multiple run and final reporters

### PR DESCRIPTION
**Pull request by insomniacslk**
This change adds support for final reporters, and raises the number of
run reporters to more than one. It also fixes a logic bug where the run
reporter was run at the end of the job instead of the end of the run.

The user can now add up several run and final reporters:
```
        "RunReporters": [
            {
                "Name": "TargetSuccess",
                "Parameters": {
                    "SuccessExpression": ">80%"
                }
            },
            {
                "Name": "noop"
            }
        ],
        "FinalReporters": [
            {
                "Name": "TargetSuccess",
                "Parameters": {
                    "AverageSuccessExpression": ">80%"
                }
            }
        ]
```

Signed-off-by: Andrea Barberio <barberio@fb.com>

### CODE REVIEW

**Comment by insomniacslk** [comment]
needs to be rebased on #31

File: pkg/jobmanager/jobmanager.go [hunk 1]
```diff
@@ -77,25 +77,38 @@ func NewJob(pr *pluginregistry.PluginRegistry, jobDescriptor string) (*job.Job,
 	if jd.RunInterval < 0 {
 		return nil, errors.New("run interval must be non-negative")
 	}
-	// TODO(insomniacslk) enable multiple reporters after the reporter
+
+	// TODO(insomniacslk) enable multiple run reporters after the reporter
 	// refactoring
 	if len(jd.Reporting.RunReporters) != 1 {
-		return nil, errors.New("exactly one reporter must be specified in a job")
+		return nil, errors.New("exactly one one run reporter must be specified in a job")
```
Comment (pmazzini) [inline pkg/jobmanager/jobmanager.go]: NIT: "one one"?

Comment (insomniacslk) [reply pkg/jobmanager/jobmanager.go]: fixed, thanks!

**Comment by marcoguerri** [comment]
I know this is not final, but if I understand correctly here were ignoring the name of the final reporter? And instead instantiating only one reporter, based on the name of the run reporter, and feeding it with the arguments of the final reporter?

**Comment by insomniacslk** [comment]
> I know this is not final, but if I understand correctly here were ignoring the name of the final reporter? And instead instantiating only one reporter, based on the name of the run reporter, and feeding it with the arguments of the final reporter?

Run and Final reporters can share the name, so I'm reusing the run reporter name when there's a final reporter, so I can build a single bundle for each plugin, and use it only if required in the job desc. But now that you mention, I may have assumed that if a final reporter is specified, the same run reporter is also specified, which is obviously wrong. I'll double-check the code tomorrow to see if I'm decoupling them appropriately

**Comment by marcoguerri** [comment]
Yes, that's I think what I meant. Just wanted to make whether this is intentional or not.

**Comment by insomniacslk** [comment]
ok, checked now. Yes, there is a bug, a final reporter is called even if not specified, as long as it's requested as run reporter. And conversely it's not called, even if specified, if not also requested as run reporter. I will separate the two logics

**Comment by insomniacslk** [comment]
Done some progress: added multiple reporters for both run and final reporting.

Need to test it better and change the database schema to:
* rename reports to final_reports
* create run_reports with the same structure of final_reports, plus the run ID

**Comment by insomniacslk** [comment]
OK, I have extended this PR to support immediately multiple run and final reporters

## example run

A job whose reporting is configured as follows:
```
    "Reporting": {
        "RunReporters": [
            {
                "Name": "TargetSuccess",
                "Parameters": {
                    "SuccessExpression": ">80%"
                }
            },
            {
                "Name": "Noop"
            }
        ],
        "FinalReporters": [
            {
                "Name": "noop"
            }
        ]
    }
```

once completed returns the following reports:
```
$ go run . status 47 | jq '.Data.Status.Report'
Requesting URL http://localhost:8080/status with requestor ID 'contestcli-http'
  with params:
    requestor: [contestcli-http]
    jobID: [47]

The server responded with status 200 OK{
  "FinalReports": [
    [
      "I did nothing at the end, all good"
    ]
  ],
  "JobID": 47,
  "ReportTime": "2020-02-07T14:21:54.156213922Z",
  "RunReports": [
    [
      {
        "AchievedSuccess": "100.00%",
        "DesiredSuccess": ">80.00%",
        "Message": "All tests pass success criteria: 100.00% > 80.00%"
      },
      "I did nothing on run #1, all good"
    ],
    [
      {
        "AchievedSuccess": "100.00%",
        "DesiredSuccess": ">80.00%",
        "Message": "All tests pass success criteria: 100.00% > 80.00%"
      },
      "I did nothing on run #2, all good"
    ],
    [
      {
        "AchievedSuccess": "100.00%",
        "DesiredSuccess": ">80.00%",
        "Message": "All tests pass success criteria: 100.00% > 80.00%"
      },
      "I did nothing on run #3, all good"
    ]
  ],
  "Success": true
}
```

## what's missing

* database schema changes as detailed above
* require at least one reporter, either run or final reporter (currently we require one run reporter and no final reporter)

**Comment by insomniacslk** [comment]
Alright, this should be functionally correct now. I expect unit/integ tests to fail because I haven't yet adjusted them to the new API, but it can be already reviewed, CC @marcoguerri @pmazzini .

Used the same job description as in my previous comment. The status is

```
$ go run . status 30 | jq 
Requesting URL http://localhost:8080/status with requestor ID 'contestcli-http'
  with params:
    requestor: [contestcli-http]
    jobID: [30]

The server responded with status 200 OK{
  "ServerID": "barberio-ubuntu-9QCS5S2",
  "Type": "ResponseTypeStatus",
  "Data": {
    "Status": {
      "EndTime": "0001-01-01T00:00:00Z",
      "JobReport": {
        "FinalReports": [
          {
            "Data": "I did nothing at the end, all good",
            "ReportTime": "2020-02-07T21:20:31Z",
            "Success": true
          }
        ],
        "JobID": 30,
        "RunReports": [
          [
            {
              "Data": {
                "AchievedSuccess": "100.00%",
                "DesiredSuccess": ">80.00%",
                "Message": "All tests pass success criteria: 100.00% > 80.00%"
              },
              "ReportTime": "2020-02-07T21:20:25Z",
              "Success": true
            },
            {
              "Data": "I did nothing on run #1, all good",
              "ReportTime": "2020-02-07T21:20:25Z",
              "Success": true
            }
          ],
          [
            {
              "Data": {
                "AchievedSuccess": "100.00%",
                "DesiredSuccess": ">80.00%",
                "Message": "All tests pass success criteria: 100.00% > 80.00%"
              },
              "ReportTime": "2020-02-07T21:20:28Z",
              "Success": true
            },
            {
              "Data": "I did nothing on run #2, all good",
              "ReportTime": "2020-02-07T21:20:28Z",
              "Success": true
            }
          ],
          [
            {
              "Data": {
                "AchievedSuccess": "100.00%",
                "DesiredSuccess": ">80.00%",
                "Message": "All tests pass success criteria: 100.00% > 80.00%"
              },
              "ReportTime": "2020-02-07T21:20:31Z",
              "Success": true
            },
            {
              "Data": "I did nothing on run #3, all good",
              "ReportTime": "2020-02-07T21:20:31Z",
              "Success": true
            }
          ]
        ]
      },
      "Name": "test job",
      "StartTime": "2020-02-07T21:20:25Z",
      "State": "",
      "TestStatus": [
        {
          "TestName": "Literal test",
          "TestStepStatus": [
            {
              "Events": [],
              "TargetStatus": [
                {
                  "Error": null,
                  "InTime": "2020-02-07T21:20:31Z",
                  "OutTime": "2020-02-07T21:20:31Z",
                  "Target": {
                    "FQDN": "",
                    "ID": "1234",
                    "Name": "example.org"
                  }
                }
              ],
              "TestStepLabel": "cmd",
              "TestStepName": "Cmd"
            }
          ]
        }
      ]
    }
  },
  "Error": null
}
```

**Comment by insomniacslk** [comment]
there are still a couple of flaky integration tests that I am investigating

**Comment by marcoguerri** [comment]
First round of review for the in-memory integration tests, these should be straightforward to fix. I expect DB integration tests to be more problematic, I am looking into those now.

1) `TestJobManagerSuite/TestJobManagerJobCrash`:
We used to have `Report{ JobReport }`, now we have `JobReport{ RunReports, FinalReports}`. `TestJobManagerJobCrash` (which tests a crashing job), asserted that `JobReport` would be nil. Now it asserts that `RunReports` is 1, but it should instead assert 0, we don't run none of the reporting if the job crashes.

2) `TestJobManagerSuite/TestJobManagerJobStartSingle`:
This test is still making assertions on the old job report structure (for which `JobReport` inside `Report` would be nil, if no report exists for a given JobID). t should be changed as follows:
```
-       assert.Equal(suite.T(), &job.JobReport{JobID: 2}, jobReport)
+       require.Nil(suite.T(), jobReport)
```

**Comment by marcoguerri** [comment]
As for the `TestJobManagerSuite` which runs with the database backend, there is an inconsistency between the database API and the in-memory layer API, and this surfaces in `TestJobManagerSuite/TestJobManagerJobStartSingle`. To make it pass with database backend we would need to change the test with:

```
-       assert.Equal(suite.T(), &job.JobReport{JobID: 2}, jobReport)
+       require.NotNil(suite.T(), jobReport)
+       require.Equal(suite.T(), 0, len(jobReport.RunReports))
+       require.Equal(suite.T(), 0, len(jobReport.FinalReports))
``

Which is not consistent with what specified above.

The other broader problem is that the database is reset after every test,  via `Reset` function `https://github.com/facebookincubator/contest/blob/master/plugins/storage/rdbms/init.go#L61` to preserve a clean state. I'll rewrite it in a transactional way to avoid these kind of situations.

**Comment by insomniacslk** [comment]
> As for the TestJobManagerSuite which runs with the database backend, there is an inconsistency between the database API and the in-memory layer

Is this showing up because the test calls `GetJobReport` directly? I assumed this was not an issue since the higher level storage API always returns nil if the error is not nil (and I couldn't wrap my head around about why this was not the case).

Also - the test was not "flaky" as I initially thought, it was just that I didn't recognize that one failure was coming from the memory storage and (after making the change) the other error was coming from the RDBMS storage. We should probably make this more obvious in the test output

**Comment by insomniacslk** [comment]
Discussed in person, and a `nil` report is a perfectly valid condition, I have aligned the rdbms code to behave like the in-memory code

**Comment by insomniacslk** [comment]
Status after the end of a job:
```
$ go run . status 31 | jq 
Requesting URL http://localhost:8080/status with requestor ID 'contestcli-http'
  with params:
    requestor: [contestcli-http]
    jobID: [31]

The server responded with status 200 OK{
  "ServerID": "barberio-ubuntu-9QCS5S2",
  "Type": "ResponseTypeStatus",
  "Data": {
    "Status": {
      "EndTime": "0001-01-01T00:00:00Z",
      "JobReport": {
        "FinalReports": [
          {
            "Data": "I did nothing at the end, all good",
            "ReportTime": "2020-02-10T16:36:49Z",
            "Success": true
          }
        ],
        "JobID": 31,
        "RunReports": [
          [
            {
              "Data": {
                "AchievedSuccess": "100.00%",
                "DesiredSuccess": ">80.00%",
                "Message": "All tests pass success criteria: 100.00% > 80.00%"
              },
              "ReportTime": "2020-02-10T16:36:43Z",
              "Success": true
            },
            {
              "Data": "I did nothing on run #1, all good",
              "ReportTime": "2020-02-10T16:36:43Z",
              "Success": true
            }
          ],
          [
            {
              "Data": {
                "AchievedSuccess": "100.00%",
                "DesiredSuccess": ">80.00%",
                "Message": "All tests pass success criteria: 100.00% > 80.00%"
              },
              "ReportTime": "2020-02-10T16:36:46Z",
              "Success": true
            },
            {
              "Data": "I did nothing on run #2, all good",
              "ReportTime": "2020-02-10T16:36:46Z",
              "Success": true
            }
          ],
          [
            {
              "Data": {
                "AchievedSuccess": "100.00%",
                "DesiredSuccess": ">80.00%",
                "Message": "All tests pass success criteria: 100.00% > 80.00%"
              },
              "ReportTime": "2020-02-10T16:36:49Z",
              "Success": true
            },
            {
              "Data": "I did nothing on run #3, all good",
              "ReportTime": "2020-02-10T16:36:49Z",
              "Success": true
            }
          ]
        ]
      },
      "Name": "test job",
      "StartTime": "2020-02-10T16:36:43Z",
      "State": "",
      "TestStatus": [
        {
          "TestName": "Literal test",
          "TestStepStatus": [
            {
              "Events": [],
              "TargetStatus": [
                {
                  "Error": null,
                  "InTime": "2020-02-10T16:36:49Z",
                  "OutTime": "2020-02-10T16:36:49Z",
                  "Target": {
                    "FQDN": "",
                    "ID": "1234",
                    "Name": "example.org"
                  }
                }
              ],
              "TestStepLabel": "cmd",
              "TestStepName": "Cmd"
            }
          ]
        }
      ]
    }
  },
  "Error": null
}
```

**Comment by marcoguerri** [comment]
Looks good to go.

**Pull request merged**